lib/spack/spack/architecture.py

```diff
@@ -329,23 +329,27 @@ def __repr__(self):
     def __str__(self):
         return self.name
 
-    def _cmp_key(self):
-        t_keys = ''.join(str(t._cmp_key()) for t in
-                         sorted(self.targets.values()))
-        o_keys = ''.join(str(o._cmp_key()) for o in
-                         sorted(self.operating_sys.values()))
-        return (self.name,
-                self.default,
-                self.front_end,
-                self.back_end,
-                self.default_os,
-                self.front_os,
-                self.back_os,
-                t_keys,
-                o_keys)
-
-
-@key_ordering
+    def _cmp_iter(self):
+        yield self.name
```

Comment:
All yielded values need to `or` to something sane for python3 comparisons.

```suggestion
        yield self.name or ''
```
etc for all compared values.

Reply:
I worked the None comparison semantics into lazy lexicographic ordering, so this is actually no longer necessary.  I think this makes our comparisons way easier to write.